python/make_datacards.py

```diff
@@ -58,43 +61,98 @@
+                card.write("shapes signal * %s %s:signal%s\n"%(finame,wsname,catExtS))
+                card.write("shapes background * %s %s:roomultipdf%s\n"%(finame,wsname,catExtB))
+                card.write("------------\n")
+                # Observation (taken directly from RooDataSet data_obs)
+                card.write("bin ch%d\n"%(binidx))
                 card.write("observation -1\n")
                 card.write("------------\n")  
-                card.write("bin %s %s\n"%(d,d))
+                # Rates (background is left freely floating)
+                card.write("bin ch%d ch%d\n"%(binidx,binidx))
                 card.write("process signal background\n")
                 card.write("process 0 1\n")
-                # Freely floating background normalization
-                card.write("rate %f 1\n"%(nSig))
+                card.write("rate %.3f 1\n"%(nSig))
                 card.write("------------\n")  
                 # Systematics
-                card.write("lumi lnN 1.016 -\n") #2.5% uncertainty on signal (fully correlated)
-                card.write("sig_normsyst lnN 1.05 -\n") #10% example systematic uncertainty on signal (fully correlated)
-                card.write("sig_mcstat lnN %f -\n"%(1.0+mcstatunc)) #MC stat. uncertainty
-                card.write("\n")
-                card.write("pdf_index_%s_%s_%s discrete\n"%(d,ext,y))
+                card.write("lumi lnN 1.016 -\n") # Integrated luminosity uncertainty on signal (fully correlated)
+                card.write("sig_triggersf lnN %.3f -\n"%(1.0+triggersyst)) # Systematic uncertainty on signal from trigger (fully correlated)
+                card.write("sig_muonsf lnN %.3f -\n"%(1.0+muonsfsyst)) # Systematic uncertainty on signal from muon RECO, ID, isolation (fully correlated)
+                card.write("sig_muonsel lnN %.3f -\n"%(1.0+muonselsyst)) # Systematic uncertainty on signal from muon additional selection (fully correlated)
+                card.write("sig_btagsf lnN %.3f -\n"%(1.0+btagsyst)) # Systematic uncertainty on signal from b-tagging (fully correlated)
+                card.write("sig_jec_ch%d lnN %.3f -\n"%(binidx,1.0+jecsyst)) # Systematic uncertainty on signal from JES (uncorrelated)
+                card.write("sig_mcstat_ch%d lnN %.3f -\n"%(binidx,1.0+mcstatunc)) # MC stat. uncertainty (uncorrelated)
+                card.write("mean param %.3f %.3f\n"%(mean,sigma)) # Shape systematic on dimuon mass mean value
+                card.write("pdf_index_%d discrete\n"%(binidx)) # For discrete profiling
                 card.close()
+                
+                # text2workspace for individual cards:
+                os.system("text2workspace.py %s/card_ch%d_%s_M%s_%s.txt -m %s"%(outDir,binidx,s,m,y,m))
+
+            # Combine cards:
+            if len(dNames)>1:
+                os.system("combineCards.py %s/card_ch1_%s_M%s_%s.txt %s/card_ch2_%s_M%s_%s.txt > %s/card_combined_%s_M%s_%s.txt"%(outDir,s,m,y,outDir,s,m,y,outDir,s,m,y))
```

Comment:
Is the plan to have separate datacards for different years?

Reply:
Not as of today, but I went once more with flexibility, in case it'll be needed in the future.
By default, all years together are considered.